docs/extensibility/custom-component.md

```diff
@@ -0,0 +1,146 @@
+
+:::code source="snippets/MailDevResource/MailKit.Client/MailKitClientSettings.cs":::
+
+The preceding code defines the `MailKitClientSettings` class with:
+
+- `ConnectionString` property that represents the connection string to the SMTP server.
+- `Credentials` property that represents the credentials to authenticate with the SMTP server.
+- `DisableHealthChecks` property that determines whether health checks are enabled.
+- `DisableTracing` property that determines whether tracing is enabled.
+- `DisableMetrics` property that determines whether metrics are enabled.
+
+## Expose component wrapper functionality
+
+The goal of .NET Aspire components is to expose the underlying client library to consumers through dependency injection. In the case of MailKit and for this example, the `SmtpClient` class is what you want to expose. You're not wrapping any functionality, but rather mapping configuration settings to an `SmtpClient` class. It's common to expose both standard and keyed-service registrations for components. Standard registrations are used when there's only one instance of a service, and keyed-service registrations are used when there are multiple instances of a service. Sometimes this can be achieved by using a factory pattern. Add the following code to the `MailKit.Client` project in a file named _MailKitClientFactory.cs_:
+
+:::code source="snippets/MailDevResource/MailKit.Client/MailKitClientFactory.cs":::
+
+The `MailKitClientFactory` class is a factory that creates an `ISmtpClient` instance based on the configuration settings. It's responsible for returning an `ISmtpClient` implementation that's connected to a configured SMTP server and optionally authenticated. Next, you need to expose the functionality for the consumers to register this factory with the dependency injection container. Add the following code to the `MailKit.Client` project in a file named MailKitClientServiceCollectionExtensions.cs_:
+
+:::code source="snippets/MailDevResource/MailKit.Client/MailKitClientServiceCollectionExtensions.cs":::
+
+The preceding code adds two extension methods on the `IHostApplicationBuilder` type, one for the standard registration of MailKit and another for keyed-registration of MailKit. Both extensions ultimately rely on the private `AddMailKitClient` method to register the `MailKitClientFactory` with the dependency injection container as a scoped-life service. The reason for registering the `MailKitClientFactory` as a scoped service is due to the fact that the connection (and authentication) operations are considered expensive and should be reused within the same scope where possible. In other words, for a single request, the same `ISmtpClient` instance should be used.
```

Comment:
> In other words, for a single request, the same `ISmtpClient` instance should be used.

The curren code doesn't really support this well. `MailKitClientFactory` is registered as a scoped service. If you get it from 2 different sections of code running on the same request, you will get the same instance of the factory.

But those 2 different sections of code will need to call `GetSmtpClientAsync`, which returns a new SmtpClient each time. Which means it won't be caching the SmtpClient across requests.